dotnet/src/Connectors/Connectors.Memory.Pinecone/PineconeVectorRecordStore.cs

```diff
@@ -0,0 +1,234 @@
+        var mapperOptions = new StorageToDataModelMapperOptions { IncludeVectors = options?.IncludeVectors ?? false };
+
+        var index = await this.GetIndexAsync(collectionName, cancellationToken).ConfigureAwait(false);
+
+        var results = await RunOperationAsync(
+            collectionName,
+            UpsertOperationName,
+            () => index.Fetch(keys, indexNamespace, cancellationToken)).ConfigureAwait(false);
+
+        foreach (var result in results.Values)
+        {
+            yield return this._mapper.MapFromStorageToDataModel(result, mapperOptions);
+        }
+    }
+
+    /// <inheritdoc />
+    public Task DeleteAsync(string key, DeleteRecordOptions? options = default, CancellationToken cancellationToken = default)
+    {
+        Verify.NotNullOrWhiteSpace(key);
+
+        return this.DeleteBatchAsync([key], options, cancellationToken);
+    }
+
+    /// <inheritdoc />
+    public async Task DeleteBatchAsync(IEnumerable<string> keys, DeleteRecordOptions? options = default, CancellationToken cancellationToken = default)
+    {
+        Verify.NotNull(keys);
+
+        var collectionName = this.ChooseCollectionName(options?.CollectionName);
+        var indexNamespace = this.GetIndexNamespace();
+
+        var index = await this.GetIndexAsync(collectionName, cancellationToken).ConfigureAwait(false);
+
+        await RunOperationAsync(
+            collectionName,
+            DeleteOperationName,
+            () => index.Delete(keys, indexNamespace, cancellationToken)).ConfigureAwait(false);
+    }
+
+    /// <inheritdoc />
+    public async Task<string> UpsertAsync(TRecord record, UpsertRecordOptions? options = default, CancellationToken cancellationToken = default)
+    {
+        Verify.NotNull(record);
+
+        var collectionName = this.ChooseCollectionName(options?.CollectionName);
+        var indexNamespace = this.GetIndexNamespace();
+
+        var index = await this.GetIndexAsync(collectionName, cancellationToken).ConfigureAwait(false);
+
+        var vector = this._mapper.MapFromDataToStorageModel(record);
+
+        await RunOperationAsync(
+            collectionName,
+            UpsertOperationName,
+            () => index.Upsert([vector], indexNamespace, cancellationToken)).ConfigureAwait(false);
+
+        return vector.Id;
+    }
+
+    /// <inheritdoc />
+    public async IAsyncEnumerable<string> UpsertBatchAsync(
+        IEnumerable<TRecord> records,
+        UpsertRecordOptions? options = default,
+        [EnumeratorCancellation] CancellationToken cancellationToken = default)
+    {
+        Verify.NotNull(records);
+
+        var collectionName = this.ChooseCollectionName(options?.CollectionName);
+        var indexNamespace = this.GetIndexNamespace();
+
+        var index = await this.GetIndexAsync(collectionName, cancellationToken).ConfigureAwait(false);
+
+        var vectors = records.Select(this._mapper.MapFromDataToStorageModel).ToList();
```

Comment:
Can you catch exceptions coming from the mappers too and wrap them in a `VectorStoreRecordMappingException`?
The idea is that customers can catch one exception type for all mapping exceptions regardless of the vector store implementation, when using the abstraction.